nvflare/app_common/ccwf/client_controller_executor.py

```diff
@@ -0,0 +1,382 @@
+                    return make_reply(ReturnCode.OK)
+
+                self.controller = self.initialize_controller(controller_id, fl_ctx)
+                self.log_info(fl_ctx, f"Starting control flow {self.controller.name}")
+
+                try:
+                    res = self.controller.control_flow(abort_signal, fl_ctx)
+                except Exception as e:
+                    error_msg = f"{controller_id} control_flow exception: {secure_format_exception(e)}"
+                    self.log_error(fl_ctx, error_msg)
+                    self.system_panic(error_msg, fl_ctx)
+
+                if abort_signal.triggered:
+                    return make_reply(ReturnCode.TASK_ABORTED)
+
+                if hasattr(self.controller, "persistor"):
+                    self.broadcast_final_result(self.controller.persistor.load(fl_ctx), fl_ctx)
+
+                self.controller.stop_controller(fl_ctx)
+
+                self.log_info(fl_ctx, f"Finished control flow {self.controller.name}")
+
+                self.update_status(action=f"finished_{controller_id}", error=None, all_done=True)
+
+            self.update_status(action=f"finished_start_task", error=None, all_done=True)
+
+            return make_reply(ReturnCode.OK)
+
+        elif task_name == self.report_final_result_task_name:
+            return self._process_final_result(shareable, fl_ctx)
+
+        else:
+            self.log_error(fl_ctx, f"Could not handle task: {task_name}")
+            return make_reply(ReturnCode.TASK_UNKNOWN)
+
+    def _get_status_report(self):
+        with self.status_lock:
+            status = self.current_status
+            must_report = False
+            if status.error:
+                must_report = True
+            elif status.timestamp:
+                must_report = True
+
+            if not must_report:
+                return None
+
+            # do status report
+            report = copy.copy(status)
+            return report
+
+    def update_status(self, last_round=None, action=None, error=None, all_done=False):
+        with self.status_lock:
+            status = self.current_status
+            status.timestamp = time.time()
+            if all_done:
+                # once marked all_done, always all_done!
+                status.all_done = True
+            if error:
+                status.error = error
+            if action:
+                status.action = action
+            if status.last_round is None:
+                status.last_round = last_round
+            elif last_round is not None and last_round > status.last_round:
+                status.last_round = last_round
+
+            status_dict = status.to_dict()
+            self.logger.info(f"updated my last status: {status_dict}")
+
+    def _process_final_result(self, request: Shareable, fl_ctx: FLContext) -> Shareable:
+        peer_ctx = fl_ctx.get_peer_context()
+        assert isinstance(peer_ctx, FLContext)
```

Comment:
assert usually just used in test code

Reply:
Good point, will remove in this PR, and can eventually remove from ClientSideController